src/helpers/unstable_mapSliceZone.ts

```diff
@@ -0,0 +1,266 @@
+import { Slice } from "../types/value/slice";
+
+/**
+ * Convert a value to a lazyily loaded module. This is useful when using
+ * functions like `() => import("...")`.
+ */
+type LazyModule<T> = () => Promise<T | { default: T }>;
+
+/**
+ * Returns the type of a `SliceLike` type.
+ *
+ * @typeParam Slice - The Slice from which the type will be extracted.
+ */
+type ExtractSliceType<Slice extends SliceLike> = Slice extends SliceLikeRestV2
+	? Slice["slice_type"]
+	: Slice extends SliceLikeGraphQL
+	? Slice["type"]
+	: never;
+
+/**
+ * The minimum required properties to represent a Prismic Slice from the Prismic
+ * Rest API V2 for the `<SliceZone>` component.
+ *
+ * If using Prismic's Rest API V2, use the `Slice` export from
+ * `@prismicio/types` for a full interface.
+ *
+ * @typeParam SliceType - Type name of the Slice.
+ */
+export type SliceLikeRestV2<SliceType extends string = string> = {
+	slice_type: Slice<SliceType>["slice_type"];
+	id?: string;
+};
+
+/**
+ * The minimum required properties to represent a Prismic Slice from the Prismic
+ * GraphQL API for the `<SliceZone>` component.
+ *
+ * @typeParam SliceType - Type name of the Slice.
+ */
+export type SliceLikeGraphQL<SliceType extends string = string> = {
+	type: Slice<SliceType>["slice_type"];
+};
+
+/**
+ * The minimum required properties to represent a Prismic Slice for the
+ * `<SliceZone>` component.
+ *
+ * If using Prismic's Rest API V2, use the `Slice` export from
+ * `@prismicio/types` for a full interface.
+ *
+ * @typeParam SliceType - Type name of the Slice.
+ */
+export type SliceLike<SliceType extends string = string> =
+	| SliceLikeRestV2<SliceType>
+	| SliceLikeGraphQL<SliceType>;
+
+/**
+ * A looser version of the `SliceZone` type from `@prismicio/types` using
+ * `SliceLike`.
+ *
+ * If using Prismic's Rest API V2, use the `SliceZone` export from
+ * `@prismicio/types` for the full type.
+ *
+ * @typeParam TSlice - The type(s) of a Slice in the Slice Zone.
+ */
+export type SliceZoneLike<TSlice extends SliceLike = SliceLike> =
+	readonly TSlice[];
+
+/**
+ * A set of properties that identify a Slice as having been mapped.
+ * `<SliceZone>` uses these properties internally to determine how props are
+ * provided to a Slice's component.
+ */
+type MappedSliceLike = {
+	/**
+	 * If `true`, this Slice has been modified from its original value using a
+	 * mapper.
+	 */
+	__mapped: true;
```

Comment:
Can we `@internal` it so it's more likely to be hidden from intellisense?

Reply:
I'm not sure if we should mark it as `@internal`. We need to expose it so consumers of the Slice Zone can use it. It is a public way to identify if a Slice has been mapped.

We initially only plan to use it within `<SliceZone>` so we can determine how to pass a component its props, but there may be other use cases for mapping a Slice Zone.